### ISSUE TITLE
QFUF delete-then-set stmt issue.

### ISSUE THREAD

**Issue by MFaisalZaki** [pyPMT/pyPMT#15]
QFUF fails to solve domains with a delete-then-set stmt such as the one found in the rover domain.

Here is an effect from one of the actions:
```
(:action communicate_image_data
 :parameters (?r - rover ?l - lander ?o - objective ?m - mode ?x - waypoint ?y - waypoint)
 :precondition (and (at ?r ?x)(at_lander ?l ?y)(have_image ?r ?o ?m)(visible ?x ?y)(available ?r)(channel_free ?l)
            )
 :effect (and (not (available ?r))(not (channel_free ?l))(channel_free ?l)(communicated_image_data ?o ?m)(available ?r)
          )
)
```

The `available` and `channel_free` are cleared and set again. The problem is that those predicates are used as preconditions, which makes the formula unsatisfiable. 

One simple solution is to detect and remove such a pattern from the effects. The grounded version removes them from both preconditions and effects. For us the solver will assign the proper precondition values.

**Comment by MFaisalZaki** [comment]
It is now fixed in [750305e](https://github.com/pyPMT/pyPMT/commit/750305eefe17d54d2460bb417a6a5cac1f2c04b0). 

@JoanEspasa, can you review the fix? If it is okay with you, can we close the issue?

**Comment by JoanEspasa** [comment]
Added some comments in the commit. Let me know what you think :)

**Comment by JoanEspasa** [comment]
The idea is to follow the same steps that FD does: https://www.fast-downward.org/ForDevelopers/ProblemTransformations

That is, when finding an action that removes and adds a fluent, we remove the delete effect. We should also document this in the README. To avoid duplicating code, it might be the best course of action to create a UP compiler to remove this and execute it before it is passed to any encoder.

**Comment by MFaisalZaki** [comment]
This issue has struck again. This time, when we add a quality metric like the oversubscription metric, such a pattern is not removed during grounding. For now, I have moved the delete-then-set function into `encoders.utilities` but a cleaner solution is to have a Compiler for such patterns. 

current fix in [daeb48f](https://github.com/pyPMT/pyPMT/commit/daeb48fc0685b294690fba7773f6961a525b1c90)

**Comment by JoanEspasa** [comment]
Yep, a compiler in UP land will ensure you never have to worry about that in pyPMT :)

**Comment by JoanEspasa** [comment]
@MFaisalZaki do you have a list of the domains that suffer from this issue?

**Comment by MFaisalZaki** [comment]
Unfortunately, I don't have a list. Yet, the one I'm sure of is the rover domain.

**Comment by JoanEspasa** [comment]
Last commit [#1ac3153](https://github.com/pyPMT/pyPMT/commit/1ac3153d147971d4965be6c3df0eeb14a1989dbd) generalises it for any problem. Needs proper testing.
